### PR TITLE
Cascading deletes

### PR DESCRIPTION
**Pull request by floere**
I've replaced the existing foreign_key constraint with ones that do cascading deletes. This enables us to remove a pod and all related entries in other tables are removed.
Even if we never remove any entries, this is the way to go :)

Put here as a discussion basis.

### CODE REVIEW

**Comment by segiddins** [comment]
Look forward to reviewing this when I get a chance!

**Comment by floere** [comment]
Great :) I am mainly interested if all cascading deletes are correct, or whether now rows would be deleted which shouldn't.

File: lib/migrate.rb [hunk 1]
```diff
@@ -3,6 +3,17 @@
 # run all trunk migrations first, then all others.
 #
 
+# Helper method
+#
+def foreign_key_delete_cascade source_table, target_table, foreign_key
+  <<-SQL
+    ALTER TABLE #{source_table} DROP CONSTRAINT #{source_table}_#{foreign_key}_fkey;
```
Comment (floere) [inline lib/migrate.rb]: We could probably also do this via Sequel, but after reading the docs it seemed that it's clearer this way.

**Comment by floere** [comment]
@segiddins @alloy This PR blocks all progress on Strata and Humus. We depend on you to make this ~~country~~ infrastructure great!

File: lib/migrate.rb (hunk 1 above)
Comment (segiddins) [inline lib/migrate.rb]: Do we need to add escaping?

**Comment by segiddins** [comment]
Seems good?

**Comment by floere** [comment]
Ok, it's coming together.
It's now able to load a specific snapshot for tests. Then it can load the pre-migration snapshot into the test DB. Finally, the test checks that deletion fails, and after the migration it does not fail anymore.

File: lib/migrate.rb (hunk 1 above)
Comment (floere) [inline lib/migrate.rb]: Good question – I think we can keep it this way, in this case.

**Comment by floere** [comment]
I've added tests for all pod delete cases.

**Comment by alloy** [comment]
Awesome! :+1:

**Comment by floere** [comment]
Thanks. It's not live yet.

It makes use of the new prod database snapshot system to load a snapshot into the DB. Then it runs this migration, and then tests against the state after the migration.

**Pull request merged**